### PR TITLE
Simplify dict validation

### PR DESCRIPTION
**Pull request by sjakobi**

### CODE REVIEW

**Comment by sjakobi** [comment]
Quite frankly I'm not 100% sure that this change doesn't inadvertedly break something.

@keleshev, @skorokithakis: Could you please take a close look?

**Comment by sjakobi** [comment]
In case this change does in fact break something, this would be a great occasion to improve the test suite!

**Comment by sjakobi** [comment]
With regards to naming, maybe `valid` or `valid_data` would be even better than `validated`?

**Comment by sjakobi** [comment]
I have removed the variable renaming stuff from this PR, so it's a bit easier too see what this change does.

File: schema.py [hunk 1]
```diff
@@ -126,21 +123,10 @@ def validate(self, data):
...
-                            x = _x
-                            raise
-                        else:
-                            coverage.add(skey)
-                            valid = True
-                            break
-                if valid:
-                    new[nkey] = nvalue
-                elif skey is not None:
-                    if x is not None:
```
Comment (sjakobi) [inline schema.py]: `x` is always `None` at this point:
* It has been initialized as `None`.
* When a new value is assigned to `x` in line 132, it is followed directly by `raise` with the effect that this if-statement is never reached.

File: schema.py [hunk 2]
```diff
@@ -126,21 +123,10 @@ def validate(self, data):
...
-                        raise SchemaError(['Invalid value for key %r' % key] +
-                                          x.autos, [e] + x.errors)
```
Comment (sjakobi) [inline schema.py]: Concluding from my comment above, we can assert that this raise-statement is dead code.

Removing it cannot possibly break anything.

I have checked that the code affected by this change is essentially the same as in [0.3.1](https://github.com/keleshev/schema/blob/0.3.1/schema.py#L107-L138) which makes it unlikely that the change will uncover any problems in the wild.

File: schema.py [hunk 3]
```diff
@@ -126,21 +123,10 @@ def validate(self, data):
...
+                        nvalue = Schema(svalue, error=e).validate(value)
```
Comment (skorokithakis) [inline schema.py]: Why can this not raise an exception?

Comment (skorokithakis) [inline schema.py]: Ah, you're just letting it bubble up.

**Comment by skorokithakis** [comment]
This looks plausible to me, however I'd need to step through it in the debugger to make sure it's doing what it's advertising. Hopefully I'll get to that soon, it's semi-complicated code which needs time to analyze.

**Comment by sjakobi** [comment]
Thanks for reviewing this, @skorokithakis.

I know it's a complicated change but the result is a lot simpler than the starting point!

File: schema.py (hunk 3 above)
Comment (sjakobi) [inline schema.py]: Exactly! :)

**Comment by skorokithakis** [comment]
This looks good to me, it's a very desirable change, and all tests pass, so I'm inclined to merge it. I wouldn't want to merge it before a push to PyPI, so I'm going to wait for @keleshev to make a release before merging.

**Comment by sjakobi** [comment]
Thanks for vetting this change, @skorokithakis!

**Comment by sjakobi** [comment]
I saw your recommendation for schema on [HN](https://news.ycombinator.com/item?id=10784096), @skorokithakis,  and thought you might me willing to merge this now? :)

**Comment by skorokithakis** [comment]
Sorry about the delay!

**Comment by sjakobi** [comment]
Cheers and merry christmas!

**Comment by sjakobi** [comment]
Code coverage is at 100% now, BTW! :fireworks:

**Comment by skorokithakis** [comment]
Wooo! Merry Christmas man!

**Pull request merged**